File: agent/src/ua.py
"""Outbound AWS SDK User-Agent solution attribution (#319).

Every AWS API call made by the agent carries two ABCA solution-attribution
segments in the ``User-Agent`` header:

    app/uksb-wt64nei4u6#{STACKNAME}      <- native AWS_SDK_UA_APP_ID env (no code here)
    md/uksb-wt64nei4u6#agent             <- static, baked once at construction

**The ``app/`` segment is emitted by the SDK itself.** Both botocore and the
JS v3 SDK read the ``AWS_SDK_UA_APP_ID`` environment variable natively and
render it as ``app/{value}`` (botocore ``configprovider.py`` maps it to the
``user_agent_appid`` config; the value charset *includes* ``#``, so the
``uksb-wt64nei4u6#{stack}`` form survives verbatim). CDK sets that env var on
every Lambda / AgentCore runtime / ECS container, so this module contributes
**nothing** to ``app/`` — and a customer can suppress it by setting the env
var to the empty string. (This is the key simplification over the original
``/``-separated design, which had to bypass the native field because ``/`` is
not a legal app-id character. Using ``#`` keeps it native.)

This module owns only the **static ``md/`` segment** — a stable
per-component label baked once via ``user_agent_extra`` at session/client
construction. There is intentionally no per-request trace handle and no
event/middleware machinery: connection pools are never re-pinned, and
request correlation is owned by X-Ray / structured-log request ids (#245),
not the User-Agent.

The TypeScript counterparts are ``cdk/src/handlers/shared/ua.ts`` and
``cli/src/ua.ts`` — the solution id, wire format, and sanitization rules
must stay identical across all three.
"""

from __future__ import annotations

import string
from typing import Any

# AWS solution-attribution id for ABCA. Also appears (deploy-time
# counterpart, #292) in the CloudFormation stack description in
# ``cdk/src/main.ts`` and in the TS mirrors of this module. Per-surface
# literal by design.
SOLUTION_ID = "uksb-wt64nei4u6"

# Stable per-component label: this surface IS the Python agent runtime.
COMPONENT = "agent"

# RFC 7230 token charset (the UA product-token alphabet). '#' is the
# scheme's structural separator and is deliberately NOT here, so a hostile
# component/label value cannot inject extra segments.
_ALLOWED = frozenset(string.ascii_letters + string.digits + "!$%&'*+-.^_`|~")


def sanitize_ua_value(raw: str) -> str:
    """Replace every non-UA-token char (incl. non-ASCII) with ``-``."""
    return "".join(c if c in _ALLOWED else "-" for c in raw)


def static_user_agent_extra() -> str:
    """The static ``md/`` segment baked at client/session construction.

    Always ``md/{SOLUTION_ID}#{COMPONENT}`` — the ``app/`` segment is
    contributed separately by the SDK from ``AWS_SDK_UA_APP_ID`` and is not
    this module's concern.
    """
    return f"md/{SOLUTION_ID}#{sanitize_ua_value(COMPONENT)}"


def client_config() -> Any:
    """``botocore.config.Config`` carrying the static ``md/`` segment.

    For direct ``boto3.client(...)`` call sites that don't go through a
    shared session (see ``aws_session.platform_client``). Merge-friendly:
    callers that already pass a ``Config`` should use ``.merge(...)``.
    """
    from botocore.config import Config

    return Config(user_agent_extra=static_user_agent_extra())

File: cli/src/ua.ts
/**
 *  MIT No Attribution
 *
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a copy of
 *  the Software without restriction, including without limitation the rights to
 *  use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 *  the Software, and to permit persons to whom the Software is furnished to do so.
 *
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 *  SOFTWARE.
 */

/**
 * Outbound AWS SDK User-Agent solution tracking (#319) — CLI surface.
 *
 * Every AWS API call made by `bgagent` carries:
 *
 *     app/uksb-wt64nei4u6/{STACKNAME}     (only when config has stack_name)
 *     md/uksb-wt64nei4u6#cli[#{PID}]
 *
 * CLI-local mirror of `cdk/src/handlers/shared/ua.ts` (the CLI package
 * cannot import from the CDK package — same mirroring convention as
 * `cli/src/types.ts`). Solution id, wire format, and sanitization rules
 * must stay identical; `agent/src/ua.py` is the Python counterpart.
 *
 * The component label is hardcoded (`cli`); the stack name comes from the
 * optional `stack_name` field in `~/.bgagent/config.json`. The trace handle
 * is the CLI process pid, set once at startup in `bin/bgagent.ts` and
 * appended per-request by the {@link withAbcaTrace} middleware.
 */

import { tryLoadConfig } from './config';

/**
 * AWS solution-tracking id for ABCA. Deploy-time counterpart (#292) lives in
 * the CloudFormation stack description in `cdk/src/main.ts`.
 */
export const SOLUTION_ID = 'uksb-wt64nei4u6';

/** Stable per-component label: this surface IS the bgagent CLI. */
const COMPONENT = 'cli';

/** App-id budget: 50-char value cap minus `uksb-wt64nei4u6/` (16) = 34. */
const STACK_NAME_MAX = 34;

/**
 * RFC 7230 token charset; `/` and `#` deliberately excluded (structural
 * separators of the scheme). Mirrors the CDK and Python implementations.
 */
const UA_TOKEN_SAFE = /[^A-Za-z0-9!$%&'*+\-.^_`|~]/g;

let currentTrace: string | undefined;

/** Replace every non-UA-token char (incl. non-ASCII) with `-`. */
export function sanitizeUaValue(raw: string): string {
  return raw.replace(UA_TOKEN_SAFE, '-');
}

/**
 * Client config fragment carrying the static ABCA UA segments. Spread into
 * every SDK client constructor: `new SecretsManagerClient({ region, ...abcaUserAgent() })`.
 *
 * Pair semantics (mirrors the CDK module): the `app/` segment is a
 * single-element pair so its literal `/` separators survive the SDK's
 * name-position escaping; the `md/` pair lets the SDK's own `name#value`
 * join produce the `#`.
 */
export function abcaUserAgent(): { customUserAgent: ([string] | [string, string])[] } {
  const pairs: ([string] | [string, string])[] = [];
  const stackName = tryLoadConfig()?.stack_name?.trim();
  if (stackName) {
    // Sanitize FIRST, then clip, so a replaced char can't be re-split.
    const clipped = sanitizeUaValue(stackName).slice(0, STACK_NAME_MAX);
    pairs.push([`app/${SOLUTION_ID}/${clipped}`]);
  }
  pairs.push([`md/${SOLUTION_ID}`, COMPONENT]);
  return { customUserAgent: pairs };
}

/** Set (or clear) the ambient trace handle (the CLI pid, set at startup). */
export function setAbcaTrace(handle?: string): void {
  currentTrace = handle || undefined;
}

/** Current trace handle, sanitized to UA-token-safe ASCII, or undefined. */
export function getAbcaTrace(): string | undefined {
  return currentTrace ? sanitizeUaValue(currentTrace) : undefined;
}

/** Structural view of a client middleware stack (avoids @smithy/types dep). */
interface MiddlewareStackLike {
  addRelativeTo(middleware: unknown, options: Record<string, unknown>): void;
}

/**
 * Append `#{TRACE}` to the outgoing User-Agent headers on every request by
 * splicing onto the static `md/` segment, after the SDK's own
 * `getUserAgentMiddleware` has rendered the headers. Mutates only the
 * header strings; the client and its connection pool are untouched.
 * No-ops on clients without a middleware stack (jest constructor mocks).
 */
export function withAbcaTrace<T>(client: T): T {
  const stack = (client as { middlewareStack?: MiddlewareStackLike }).middlewareStack;
  if (!stack || typeof stack.addRelativeTo !== 'function') {
    return client;
  }
  const md = `md/${SOLUTION_ID}#${COMPONENT}`;
  stack.addRelativeTo(
    (next: (args: unknown) => Promise<unknown>) => async (args: unknown) => {
      const trace = getAbcaTrace();
      const request = (args as { request?: { headers?: Record<string, string> } }).request;
      if (trace && request?.headers) {
        for (const header of ['user-agent', 'x-amz-user-agent']) {
          const value = request.headers[header];
          if (value && value.includes(md)) {
            request.headers[header] = value.replace(md, `${md}#${trace}`);
          }
        }
      }
      return next(args);
    },
    {
      name: 'abcaUaTraceMiddleware',
      relation: 'after',
      toMiddleware: 'getUserAgentMiddleware',
      override: true,
    },
  );
  return client;
}
